route-rs-packets/src/ethernet.rs

```diff
@@ -0,0 +1,169 @@
+        // Ethernet II frames must be at least the header, which is 14bytes
+        // 0                    6                    12                      14
+        // |---6 byte Dest_MAC--|---6 byte Src_MAC---|--2 Byte Payload Len---|
+        // We could support other formats for the frames, but IP sits atop Ethernet II
+
+        // If we want to have formated errors, we will need to create an object that implements
+        // Error that we can return, rather than making the error a &'static str
+        // Will create an issue, so it can be done later.
+        if frame.len() < 14 {
+            return Err("Frame is less than the minimum of 14 bytes");
+        }
+        // Can't use helper here, since we don't have the object yet :(.
+        let payload_len = u16::from_be_bytes([frame[12], frame[13]]) as usize;
+        if payload_len + 14 != frame.len() {
+            return Err("Frame has invalivd payload len field");
+        }
+
+        Ok(EthernetFrame {
+            data: frame,
+            payload_offset: 14,
+        })
+    }
+
+    pub fn dest_mac(&self) -> MacAddr {
+        let bytes = <[u8; 6]>::try_from(&self.data[0..6]).unwrap();
+        MacAddr::new(bytes)
+    }
+
+    pub fn src_mac(&self) -> MacAddr {
+        let bytes = <[u8; 6]>::try_from(&self.data[6..12]).unwrap();
+        MacAddr::new(bytes)
+    }
+
+    pub fn set_dest_mac(&mut self, mac: MacAddr) {
+        self.data[..6].copy_from_slice(&mac.bytes[..6]);
+    }
+
+    pub fn set_src_mac(&mut self, mac: MacAddr) {
+        self.data[6..12].copy_from_slice(&mac.bytes[..6]);
+    }
+
+    pub fn payload_len(&self) -> u16 {
+        u16::from_be_bytes(self.data[12..=13].try_into().unwrap())
+    }
+
+    // This gives you a cow of a slice of the payload.
+    pub fn payload(&self) -> Cow<[u8]> {
+        Cow::from(&self.data[self.payload_offset..])
+    }
+
+    pub fn set_payload(&mut self, payload: &[u8]) {
+        let payload_len = payload.len() as u16;
+        self.data.truncate(self.payload_offset - 2); // Cut off and drop the entire payload and payload_len
+        let payload_len_bytes = payload_len.to_be_bytes();
+        self.data.reserve_exact(payload_len as usize + 2); // Reserve room for payload and len field.
+        self.data.extend(payload_len_bytes.iter());
+        self.data.extend(payload);
+    }
+}
+
+impl<'packet> TryFrom<TcpSegment<'packet>> for EthernetFrame<'packet> {
+    type Error = &'static str;
+
+    fn try_from(segment: TcpSegment<'packet>) -> Result<Self, Self::Error> {
+        EthernetFrame::new(segment.data)
+    }
+}
+
+impl<'packet> TryFrom<UdpSegment<'packet>> for EthernetFrame<'packet> {
+    type Error = &'static str;
+
+    fn try_from(segment: UdpSegment<'packet>) -> Result<Self, Self::Error> {
+        EthernetFrame::new(segment.data)
+    }
+}
+
+impl<'packet> TryFrom<Ipv4Packet<'packet>> for EthernetFrame<'packet> {
+    type Error = &'static str;
+
```

Comment:
Should we have a way of encapsulating an Ipv4 packet if it didn't originally come from an ethernet frame?

Reply:
In this case, this function creates an EthernetFrame from an Ipv4 or Ipv6.  So it isn't really encapsulating per se, just a type conversion.  We could add support for other Layer 2 protocols if need be, since the try_from would fail in that caseif you tried to make an ethernet frame.

Reply:
I'm thinking of the case where say a processor/element creates an Arp Message, that now needs to be put behind an Ethernet header to be sent out over the wire. These would be separate from the TryFroms, but could be methods that take an arbitrary payload, a src/dest MAC addr, and an ethertype, then replaces the old l2 header or pushes more data onto the front of the packet as necessary